File: codeforces/count-substrings.py
def count_distinct_substrings(s):
    n = len(s)
    
    # Validate input
    if not (1 <= n <= 400000):
        return "Error: String length must be between 1 and 400000"
    if not s.islower() or not s.isalpha():
        return "Error: String must contain only lowercase English letters"
    
    # Step 1: Build suffix array (simplified version)
    suffixes = [(s[i:], i) for i in range(n)]
    suffixes.sort()  # Sort by suffix string
    
    # Step 2: Count distinct substrings using LCP
    total = n - suffixes[0][1]  # Length of first suffix
    
    # Compare adjacent suffixes
    for i in range(1, n):
        curr_suffix, curr_idx = suffixes[i]
        prev_suffix, prev_idx = suffixes[i-1]
        
        # Compute LCP (Longest Common Prefix) length
        lcp = 0
        for j in range(min(len(curr_suffix), len(prev_suffix))):
            if curr_suffix[j] != prev_suffix[j]:
                break
            lcp += 1
        
        # Add new unique characters from current suffix
        total += (n - curr_idx) - lcp
    
    return total

# Get input and print result
s = input().strip()
result = count_distinct_substrings(s)
print(result)

File: codeforces/count-substrings.cpp
#include <iostream>
#include <string>
#include <memory>

using namespace std;

// Trie node structure
struct TrieNode {
    TrieNode* children[26];  // For lowercase a-z
    TrieNode() {
        for (int i = 0; i < 26; i++) {
            children[i] = nullptr;
        }
    }
};

// Function to count distinct substrings using Trie
long long count_distinct_substrings(const string& s) {
    int n = s.length();
    
    // Validate input
    if (n < 1 || n > 400000) {
        cout << "Error: String length must be between 1 and 400000" << endl;
        return -1;
    }
    for (char c : s) {
        if (!islower(c) || !isalpha(c)) {
            cout << "Error: String must contain only lowercase English letters" << endl;
            return -1;
        }
    }
    
    // Create Trie root
    unique_ptr<TrieNode> root(new TrieNode());
    long long count = 0;  // Count of distinct substrings
    
    // Insert all substrings into Trie
    for (int i = 0; i < n; i++) {
        TrieNode* current = root.get();
        for (int j = i; j < n; j++) {
            int index = s[j] - 'a';
            if (!current->children[index]) {
                current->children[index] = new TrieNode();
                count++;  // New substring found
            }
            current = current->children[index];
        }
    }
    
    return count;
}

int main() {
    string s;
    getline(cin, s);
    
    long long result = count_distinct_substrings(s);
    if (result != -1) {
        cout << result << endl;
    }
    
    return 0;
}
